lib/index.ts

```diff
@@ -58,75 +71,161 @@ export async function buildDepTree(
-
-  return tree;
-}
-
-function buildDependencyTree(
-  manifestFile: PaketDependencies,
-  lockFile: PaketLock,
-  includeDev: boolean = false,
-) {
   const depTree = {
     dependencies: {},
-    name: '',
+    name: defaultManifestFileName,
     version: '',
   } as DepTree;
 
-  const dependencies: { [dep: string]: {} } = {};
+  const dependenciesMap: Map<string, FlatDependency> = new Map();
+
+  collectRootDeps(manifestFile, dependenciesMap);
+  collectResolvedDeps(lockFile, dependenciesMap);
+
+  for (const dep of dependenciesMap.values()) {
+    if (dep.root) {
+      calculateReferences(dep, dependenciesMap);
+    }
+  }
+
+  for (const dep of dependenciesMap.values()) {
+    if (dep.root && (includeDev || dep.depType === DepType.prod)) {
+      if (strict && !dep.resolved) {
+        throw new OutOfSyncError(dep.name);
+      }
+
+      depTree.dependencies[dep.name] = buildTreeFromList(dep, dependenciesMap);
 
+      if (!dep.resolved) {
+        depTree.dependencies[dep.name].missingLockFileEntry = true;
+      }
+    }
+  }
+
+  const frequentSubTree = buildFrequentDepsSubtree(dependenciesMap);
+
+  if (Object.keys(frequentSubTree.dependencies).length) {
+    depTree.dependencies[frequentSubTree.name] = frequentSubTree;
+  }
+
+  return depTree;
+}
+
+function collectRootDeps(manifestFile: PaketDependencies, dependenciesMap: Map<string, FlatDependency>) {
   for (const group of manifestFile) {
+    const isDev = DEV_GROUPS.includes((group.name || '').toLowerCase());
+
     for (const dep of group.dependencies) {
-      const current = dep as NugetDependency;
-      if (current.name) {
-        dependencies[current.name] = {
-          name: current.name,
-        };
+      if (dep.source !== 'nuget') {
+        continue;
+      }
+
+      const nugetDep = dep as NugetDependency;
+
+      if (!dependenciesMap.has(nugetDep.name.toLowerCase())) {
+        dependenciesMap.set(nugetDep.name.toLowerCase(), {
+          name: nugetDep.name,
+          // Will be overwritten in `collectResolvedDeps`.
+          version: nugetDep.versionRange,
+          // Will be overwritten in `collectResolvedDeps`.
+          dependencies: [],
+          depType: isDev ? DepType.dev : DepType.prod,
+          root: true,
+          refs: 1,
+          // Will be overwritten in `collectResolvedDeps`.
+          resolved: false,
+        });
       }
     }
   }
+}
 
+function collectResolvedDeps(lockFile: PaketLock, dependenciesMap: Map<string, FlatDependency>) {
   for (const group of lockFile.groups) {
     const isDev = DEV_GROUPS.includes((group.name || '').toLowerCase());
 
-    if (isDev && !includeDev) {
-      continue;
-    }
-
     for (const dep of group.dependencies) {
-      if (dependencies[dep.name]) {
-        dependencies[dep.name] = {
-          depType: isDev ? DepType.dev : DepType.prod,
-          dependencies: buildSubTree(dep.dependencies, group.dependencies, isDev),
+      if (dep.repository !== 'NUGET') {
```

Comment:
should we have a list of supported repository types? That way we can expand this more easily at some later date

Reply:
done